feat(pointpainting_fusion): add diagnostics for processing time of pointpainting

a-maumau opened the pull request:
## Description
This PR adds diagnostics for pointpainting's processing time.

If the processing time exceeds `max_allowed_processing_time_ms`, a warning topic will be published.

Consecutive delays will be also checked.
If the difference of two timestamp (current and last in-time processed timestamp) is larger than `max_acceptable_consecutive_delay_ms`, it will determine as an error state.

If the new processing time is below the `max_allowed_processing_time_ms`, it will update the timestamp of last in-time processed case.



## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?
Tested by checking the rqt_runtime_monitor.

Run rqt_runtime_monitor by
```sh
ros2 run rqt_runtime_monitor rqt_runtime_monitor
```

![Screenshot from 2025-04-04 13-56-44](https://github.com/user-attachments/assets/ced9c43f-a6a5-4e89-a2c6-cb6de67cbf84)

## Notes for reviewers

None.

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

badai-nguyen commented on perception/autoware_image_projection_based_fusion/launch/pointpainting_fusion.launch.xml:
```diff
@@ -25,6 +25,7 @@
   <arg name="ml_package_param_path" default="$(var model_path)/$(var model_name)_ml_package.param.yaml"/>
   <arg name="class_remapper_param_path" default="$(var model_path)/detection_class_remapper.param.yaml"/>
   <arg name="sync_param_path" default="$(find-pkg-share autoware_image_projection_based_fusion)/config/fusion_common.param.yaml"/>
+  <arg name="common_param_path" default="$(find-pkg-share autoware_image_projection_based_fusion)/config/pointpainting_common.param.yaml"/>
```
@a-maumau Thanks for your PR.
cc @technolojin

```suggestion
  <arg name="common_param_path" default="$(find-pkg-share autoware_image_projection_based_fusion)/config/pointpainting_diagnostic.param.yaml"/>
```
- If you dont have a parameters located in this param files, I prefer `pointpainting_diagnostic.param.yaml` 

- BTW, Currently, there are already 3 param files related to pointpainting, I think merging diagnostic param to one of them could be better? such as pointpainting.param.yaml. How do you think?

technolojin replied:
@badai-nguyen 

existing 3 parameter files are related to the detection model.
This configuration is for node implementation.

To combine parameter files, the diagnostics parameter files is named pointpainting_common.

If too many parameter files are issue, we may need to integrate the model related parameter files to 1 or 2.

a-maumau replied:
@badai-nguyen thank you for your suggestion!
For my answer for this is like https://github.com/autowarefoundation/autoware_universe/pull/10398#issuecomment-2781824805
TL;DR: I agree with your points, and also agree with @technolojin 's points.
So we should discuss this considering future development/features and file management.

badai-nguyen replied:
> So we should discuss this considering future development/features and file management.

@a-maumau @technolojin Then I think it good for now. However, IMO, we should have a further discussion about parameter file management later. Especially,  including the consolidation of other diagnostics parameters files from other package into the same directory to make it easier to manage the MRM trigger conditions.

technolojin replied:
I agree. From my perspective, this topic just triggered a common issue of the lidar detector parameter management.

For MRM trigger condition, it is discussed in the failure detection WG (incl. Horibe-san). 

- Each component/node/package shall have its own diagnostics parameter (not MRM config)
- MRM condition is configured in the system side
 
Therefore, node configuration will not be in a same place

technolojin reviewed (changes requested):
The evaluating processing time contains all the message synchronization latency.
`stop_watch_ptr_->toc("processing_time", true)`

1. It could be a very fluctuating value due to input message timings.
2. It is not actual `processing time` of the model
3. The process of image projection is done my multiple callbacks. It is hard to expect total amount of process.

Therefore, I would like to suggest to measure independent processing times, such as

- `fuse_on_single_image` loop
- `export_process` or `postprocess`

a-maumau commented:
@technolojin 
> The evaluating processing time contains all the message synchronization latency.
stop_watch_ptr_->toc("processing_time", true)

Sorry, I overlooked my typo.
The part of `stop_watch_ptr_->toc("processing_time")` was not intended, it should be `stop_watch_ptr_->toc("pointpainting_processing_time")`, which I fixed it in [c7cdf57](https://github.com/autowarefoundation/autoware_universe/pull/10397/commits/c7cdf57163c5e6e4f940b12856bec78eed9c3f6a).

So, it is now only checking the `postprocess` function, and corresponding with this change, I fixed the parameter in [b7b3257](https://github.com/autowarefoundation/autoware_universe/pull/10397/commits/b7b3257b37e3a13663e7a791925f17f923272825).

technolojin reviewed (approved):
LGTM

badai-nguyen commented:
@a-maumau cc @technolojin 
I wondering why we don't use the same processing_time here for diagnostics? which is recorded in rosbag and could be used for later debuging purpose?
https://github.com/autowarefoundation/autoware_universe/blob/e9d43e45ca477ec70e58421be59cebf867f11f5e/perception/autoware_image_projection_based_fusion/src/fusion_node.cpp#L333-L334

technolojin commented:
> @a-maumau cc @technolojin I wondering why we don't use the same processing_time here for diagnostics? which is recorded in rosbag and could be used for later debuging purpose?
> 
> https://github.com/autowarefoundation/autoware_universe/blob/e9d43e45ca477ec70e58421be59cebf867f11f5e/perception/autoware_image_projection_based_fusion/src/fusion_node.cpp#L333-L334

The implementation of the `processing_time` includes fuse_on_single_image, but it is not predictable.
* not fixed number of ROIs
* not fixed trigger timing

It is impossible to set a proper fixed processing time threshold, if we use the `processing_time`.

Therefore, in the pointpainting, implementing diagnostics only for the ml model inference part.

badai-nguyen reviewed (approved):
LGTM
